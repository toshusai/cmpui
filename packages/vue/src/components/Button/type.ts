export const ButtonVariant = {
  primary: "primary",
  secondary: "secondary",
} as const;

export type ButtonVariant = (typeof ButtonVariant)[keyof typeof ButtonVariant];

export const ButtonSize = {
  S: "S",
  M: "M",
  L: "L",
} as const;

export type ButtonSize = (typeof ButtonSize)[keyof typeof ButtonSize];

export type ButtonProps = {
  variant: ButtonVariant;
  size: ButtonSize;
  disabled: boolean;
};
