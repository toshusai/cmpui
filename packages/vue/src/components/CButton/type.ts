export const CButtonVariant = {
  primary: "primary",
  secondary: "secondary",
} as const;

export type CButtonVariant =
  (typeof CButtonVariant)[keyof typeof CButtonVariant];

export const CButtonSize = {
  S: "S",
  M: "M",
  L: "L",
} as const;

export type CButtonSize = (typeof CButtonSize)[keyof typeof CButtonSize];

export type CButtonProps = {
  variant?: CButtonVariant;
  size?: CButtonSize;
  disabled?: boolean;
};
