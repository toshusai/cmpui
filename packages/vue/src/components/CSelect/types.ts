import { VNode } from "vue";

export type SelectProps = {
  disabled?: boolean;
  value?: string;
  defaultShow?: boolean;
  label?: string;
  options?: {
    label: string | VNode;
    value: string;
  }[];
};
