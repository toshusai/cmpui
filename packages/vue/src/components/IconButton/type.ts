export const IconButtonSize = {
  S: "S",
  M: "M",
} as const;

export type IconButtonSize =
  (typeof IconButtonSize)[keyof typeof IconButtonSize];

export type IconButtonProps = {
  bordered?: boolean;
  disabled?: boolean;
  rounded?: boolean;
  selected?: boolean;
  size?: IconButtonSize;
};
