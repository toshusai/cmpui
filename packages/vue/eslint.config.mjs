"use strict";
import pluginVue from "eslint-plugin-vue";
import eslintConfigPrettier from "eslint-config-prettier";
import tseslint from "typescript-eslint";

export default [
  ...tseslint.configs.recommended,
  ...pluginVue.configs["flat/recommended"],
  {
    ignores: ["dist/*", "types/*"],
  },
  {
    rules: {
      "vue/multi-word-component-names": "off",
      "no-unused-vars": "off",
      "@typescript-eslint/no-unused-vars": ["error"],
    },
    languageOptions: {
      parserOptions: { parser: tseslint.parser, sourceType: "module" },
    },
  },
  eslintConfigPrettier,
];
