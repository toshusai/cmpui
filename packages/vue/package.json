{
  "name": "@toshusai/cmpui-vue",
  "version": "1.0.0-alpha.8",
  "description": "CmpUI is a component library designed for creating complex GUI applications in web browsers.",
  "main": "./dist/index.umd.js",
  "module": "./dist/index.es.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "import": {
        "default": "./dist/index.es.js",
        "types": "./dist/index.d.ts"
      },
      "require": {
        "default": "./dist/index.umd.js",
        "types": "./dist/index.d.ts"
      }
    }
  },
  "scripts": {
    "typecheck": "vue-tsc -p tsconfig.ci.json",
    "test:watch": "vitest",
    "test": "vitest run",
    "build": "npm run typecheck && vite build && vue-tsc",
    "storybook": "storybook dev -p 6006",
    "build-storybook": "storybook build",
    "format": "prettier --write .",
    "lint": "eslint --fix .",
    "fix": "npm run format && npm run lint"
  },
  "dependencies": {
    "@toshusai/cmpui-css": "link:../css",
    "vue": "^3.4.38"
  },
  "devDependencies": {
    "@chromatic-com/storybook": "^1.7.0",
    "@storybook/addon-essentials": "8.2.6",
    "@storybook/addon-interactions": "8.2.6",
    "@storybook/addon-links": "8.2.6",
    "@storybook/addon-mdx-gfm": "8.2.6",
    "@storybook/addon-onboarding": "8.2.6",
    "@storybook/addon-themes": "8.2.6",
    "@storybook/blocks": "8.2.6",
    "@storybook/test": "8.2.6",
    "@storybook/test-runner": "^0.17.0",
    "@storybook/theming": "8.2.6",
    "@storybook/vue3": "^8.2.9",
    "@storybook/vue3-vite": "^8.2.9",
    "@types/jest": "^29.5.11",
    "@vitejs/plugin-vue": "^5.1.2",
    "@vue/compiler-sfc": "^3.4.38",
    "concurrently": "^8.2.2",
    "eslint": "^9.9.0",
    "eslint-config-prettier": "^9.1.0",
    "eslint-plugin-vue": "^9.27.0",
    "storybook": "8.2.6",
    "typescript": "^5.2.2",
    "typescript-eslint": "^8.2.0",
    "vitest": "^1.1.0",
    "vue-tsc": "^2.0.29"
  },
  "keywords": [
    "typescript",
    "components",
    "ui",
    "gui"
  ],
  "author": "toshusai",
  "license": "MIT"
}
