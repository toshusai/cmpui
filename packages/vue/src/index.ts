import Button from "./components/Button/Button.vue";

export { Button };
