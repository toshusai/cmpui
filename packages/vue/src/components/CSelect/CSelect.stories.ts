import type { Meta, StoryObj } from "@storybook/vue3";
import CSelect from "./CSelect.vue";
import { h, ref } from "vue";

const meta = {
  component: CSelect,
} satisfies Meta<typeof CSelect>;

export default meta;
type Story = StoryObj<typeof meta>;

const html = String.raw;
const names = [
  "Alfa",
  "Bravo",
  "Charlie",
  "Delta",
  "Echo",
  "Foxtrot",
  "Golf",
  "Hotel",
  "India",
  "Juliett",
  "Kilo",
  "Lima",
  "Mike",
  "November",
  "Oscar",
  "Papa",
  "Quebec",
  "Romeo",
  "Sierra",
  "Tango",
  "Uniform",
  "Victor",
  "Whiskey",
  "X-ray",
  "Yankee",
  "Zulu",
];

export const Default: Story = {
  render: () => ({
    setup() {
      const value = ref<string>(names[8]);
      return {
        names,
        value,
      };
    },
    components: { CSelect },
    template: html`
      <CSelect
        :value="value"
        :options="names.map((name) => ({ label: name, value: name }))"
        @change="(v) => (value = v)"
      ></CSelect>
    `,
  }),
};

export const Label: Story = {
  render: () => ({
    setup() {
      const value = ref<string>(names[8]);
      return {
        names,
        value,
      };
    },
    components: { CSelect },
    template: html`
      <CSelect
        :value="value"
        :options="names.map((name) => ({ label: name, value: name }))"
        label="Label"
        @change="(v) => (value = v)"
      ></CSelect>
    `,
  }),
};

export const Disabled: Story = {
  render: () => ({
    setup() {
      const value = ref<string>(names[8]);
      return {
        names,
        value,
      };
    },
    components: { CSelect },
    template: html`
      <CSelect
        disabled
        :value="value"
        :options="names.map((name) => ({ label: name, value: name }))"
        label="Label"
        @change="(v) => (value = v)"
      ></CSelect>
    `,
  }),
};

export const Overflow: Story = {
  render: () => ({
    setup() {
      const value = ref<string>(names[8]);
      return {
        names,
        value,
      };
    },
    components: { CSelect },
    template: html`
      <CSelect
        style="width: 128px"
        :value="value"
        :options="names.map((name) => ({ label: name.repeat(16), value: name }))"
        @change="(v) => (value = v)"
      ></CSelect>
    `,
  }),
};

export const Custom: Story = {
  render: () => ({
    setup() {
      const value = ref<string>(names[8]);
      const options = names.map((name) => ({
        label: name,
        value: name,
        render: (value) => h("strong", value),
      }));
      return {
        names,
        value,
        options,
      };
    },
    components: { CSelect },
    template: html`
      <CSelect
        style="width: 128px"
        :value="value"
        :options="options"
        @change="(v) => (value = v)"
      ></CSelect>
    `,
  }),
};
