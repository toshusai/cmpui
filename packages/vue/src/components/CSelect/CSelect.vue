<script setup lang="ts">
import { computed, onUnmounted, ref, watch } from "vue";
import Popover from "../CPopover/CPopover.vue";
import CMenuList from "../CMenuList/CMenuList.vue";
import CMenuListItem from "../CMenuListItem/CMenuListItem.vue";
import { lockScroll } from "../../../../core/src/dialog/lockScroll";
import { SelectProps } from "./types";
import { useId } from "../../lib/useId";

const props = withDefaults(defineProps<SelectProps>(), {
  disabled: false,
  defaultShow: undefined,
  label: undefined,
  options: undefined,
});

const value = defineModel<string>();

const show = ref(false);
const trigger = ref<HTMLElement | null>(null);

const preview = computed(() => {
  const option = props.options?.find((o) => o.value === value.value);
  if (option) {
    if (typeof option.label === "string") return option.label;
    return option.label;
  }
  return null;
});

const handleSelect = (v: string) => {
  value.value = v;
  show.value = false;
  trigger.value?.focus();
};

let cleanUp = () => {};

watch(
  () => show.value,
  (show) => {
    if (show) {
      const cleanUpLockScroll = lockScroll();
      cleanUp = () => {
        cleanUpLockScroll();
      };
    } else {
      cleanUp();
    }
  },
);

onUnmounted(() => {
  cleanUp();
});

const enablePointerUpSelect = ref(false);

const handlePointerDown = () => {
  if (props.disabled) return;
  let firstState = show.value;
  show.value = true;

  enablePointerUpSelect.value = false;
  let distance = 0;
  const pointerMove = (e: PointerEvent) => {
    distance += Math.abs(e.movementX) + Math.abs(e.movementY);
    if (distance > 10) {
      enablePointerUpSelect.value = true;
    }
  };

  const pUp = () => {
    if (firstState) {
      show.value = false;
    }
    window.removeEventListener("pointermove", pointerMove);
    window.removeEventListener("pointerup", pUp);
  };

  window.addEventListener("pointermove", pointerMove);
  window.addEventListener("pointerup", pUp);
};

const id = useId();
const labelId = useId();
const popupId = useId();
</script>

<template>
  <button
    :id="id"
    ref="trigger"
    class="cmpui_select__root"
    :disabled="disabled"
    v-bind="$attrs"
    role="combobox"
    :aria-expanded="show"
    :aria-labelledby="labelId"
    :aria-controls="popupId"
    aria-haspopup="listbox"
    @pointerdown="handlePointerDown"
    @keydown.enter="show = true"
    @keydown.space="show = true"
  >
    <div class="cmpui_select__preview">
      <span v-if="typeof preview === 'string'">{{ preview }}</span>
      <component :is="preview" v-else />
    </div>
    <label v-if="label" :id="labelId" class="cmpui_select__label" :for="id">
      {{ label }}
    </label>
  </button>
  <Popover
    v-if="show"
    disabled-trigger-click-close
    auto-resize
    :show="show"
    :trigger="trigger"
    @close="show = false"
  >
    <CMenuList
      :id="popupId"
      :default-value="value"
      :aria-labelledby="labelId"
      role="listbox"
      @select="handleSelect"
    >
      <CMenuListItem
        v-for="option in options"
        :key="option.value"
        rounded
        style="padding-left: 24px; position: relative"
        role="option"
        :value="option.value"
        :data-value="option.value"
        @click="handleSelect(option.value)"
        @pointerup="enablePointerUpSelect ? handleSelect(option.value) : null"
        @keydown.tab="handleSelect(option.value)"
      >
        <div
          v-if="option.value === value"
          class="cmpui_select_check-icon"
          style="position: absolute; left: 8px"
        ></div>
        <component :is="option.label" v-if="typeof option.label !== 'string'" />
        <template v-else>
          {{ option.label }}
        </template>
      </CMenuListItem>
    </CMenuList>
  </Popover>
</template>
