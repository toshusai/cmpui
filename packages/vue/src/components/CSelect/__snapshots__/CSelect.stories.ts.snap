// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CSelect Default smoke-test 1`] = `
<button id="id-1"
        class="cmpui_select__root"
        role="combobox"
        aria-expanded="false"
        aria-labelledby="id-2"
        aria-controls="id-3"
        aria-haspopup="listbox"
>
  <span>
    India
  </span>
  <div class="cmpui_select__keyboard-arrow-down-icon">
  </div>
</button>
`;

exports[`components/CSelect Disabled smoke-test 1`] = `
<button id="id-7"
        class="cmpui_select__root"
        disabled
        role="combobox"
        aria-expanded="false"
        aria-labelledby="id-8"
        aria-controls="id-9"
        aria-haspopup="listbox"
>
  <span>
    India
  </span>
  <div class="cmpui_select__keyboard-arrow-down-icon">
  </div>
  <label id="id-8"
         class="cmpui_select__label"
         for="id-7"
  >
    Label
  </label>
</button>
`;

exports[`components/CSelect Label smoke-test 1`] = `
<button id="id-4"
        class="cmpui_select__root"
        role="combobox"
        aria-expanded="false"
        aria-labelledby="id-5"
        aria-controls="id-6"
        aria-haspopup="listbox"
>
  <span>
    India
  </span>
  <div class="cmpui_select__keyboard-arrow-down-icon">
  </div>
  <label id="id-5"
         class="cmpui_select__label"
         for="id-4"
  >
    Label
  </label>
</button>
`;
